### ISSUE TITLE
MusicgenMelody ignores audio conditioning (regression between 4.48 and 4.57)

### ISSUE THREAD

**Issue by audiodude** [huggingface/transformers#45647]
_(Original Note): Claude removed this line when editing, but I wanted to fully disclose that this issue was discovered and written up by Claude code_

**Update (corrected):** the regression is wider than the title suggests — it already exists in transformers **4.57.6**, the latest 4.x. So this is not a v5 regression; it broke somewhere between 4.48 and 4.57.

### System info

- `transformers` versions tested:
  - **4.48.3** — works (audio conditioning active)
  - **4.57.6** — broken (audio ignored)
  - **5.5.4** — broken (audio ignored)
- Python 3.12.13, PyTorch 2.11.0+cu130, CUDA on RTX 3080 Ti, fp16
- Model: `facebook/musicgen-melody`

### Description

`MusicgenMelodyForConditionalGeneration.generate()` ignores the audio reference (`input_features`) in transformers ≥ 4.57. Two reference audios with different chroma classes produce **byte-identical** generated audio for the same text prompt and seed. The same code in 4.48 produces meaningfully different output.

### Minimal reproducer

```python
import numpy as np, torch
from transformers import AutoProcessor, MusicgenMelodyForConditionalGeneration

device, dtype = "cuda", torch.float16
proc = AutoProcessor.from_pretrained("facebook/musicgen-melody")
model = MusicgenMelodyForConditionalGeneration.from_pretrained(
    "facebook/musicgen-melody", torch_dtype=dtype
).to(device)

sr = 32000
t = np.linspace(0, 4, sr * 4)
A  = (0.4 * np.sin(2 * np.pi * 440.00 * t)).astype(np.float32)  # A4
Eb = (0.4 * np.sin(2 * np.pi * 311.13 * t)).astype(np.float32)  # Eb4 — different chroma class

def gen(audio):
    inputs = proc(text=["jazz"], audio=audio, sampling_rate=sr,
                  padding=True, return_tensors="pt").to(device)
    inputs["input_features"] = inputs["input_features"].to(dtype)
    torch.manual_seed(42)
    return model.generate(**inputs, max_new_tokens=100,
                          do_sample=True, guidance_scale=3.0)

a, e = gen(A), gen(Eb)
diff = (a[0, 0].float() - e[0, 0].float()).abs().mean().item()
print(f"output diff (A vs Eb): {diff:.4f}")
```

### Results

| transformers | output diff (A vs Eb) | audio conditioning |
|---|---|---|
| 4.48.3 | **0.1610** | works |
| 4.57.6 | **0.0000** | ignored |
| 5.5.4  | **0.0000** | ignored |

### Where the chain breaks (per v5.5.4 tracing)

- The processor extracts chroma features correctly. `input_features` is shape `(1, N, 12)` and values differ between A and Eb (chroma abs diff ≈ 0.17).
- `_prepare_encoder_hidden_states_kwargs_for_generation` does receive `input_features` in `model_kwargs` (verified by hooking).
- The returned `encoder_hidden_states` (audio prefix concatenated with text encoder output) differs between A and Eb — mean abs diff ≈ 0.015 against mean abs ≈ 0.034, i.e. the audio-prefix portion is meaningfully different.
- But per-step logits returned by `model.generate` are byte-identical between the two audios.

So the audio conditioning reaches the encoder-hidden-states side correctly, but the decoder produces identical logits regardless — suggesting the audio prefix is not actually being attended to in the decoder. dtype was not the cause; promoting `audio_enc_to_dec_proj` to fp32 with cast hooks did not change the result.

### Bisection range

Broken: 4.57.6, 5.5.4. Last-known-good: 4.48.3. Haven't bisected within the 4.48 → 4.57 range.

### Disclosure

This regression was identified by Claude Code during a debugging session — disclosing for transparency.

**Comment by voodoovampire** [comment]
Hi, this PR updates the SDPA inference tolerances used in the MPS backend by routing "mps" through the same tolerance branch as "xpu" in the SDPA tests. On MPS, the default tolerances were too strict for fp16 and caused flaky failures even though the numerical differences were within expected fp16 ranges. All CI checks are passing; happy to adjust anything if you’d like a different tolerance strategy. I’m happy to propose a minimal PR plus a unit test that checks different input_features do not produce identical outputs for a fixed seed

**Comment by geurinmccuan345-maker** [comment]
Ok tänks bro

في السبت، 25 أبريل 2026 19:59 Travis Briggs ***@***.***> كتب:

**Comment by Rocketknight1** [comment]
cc @ebezzam @eustlb

**Comment by adityachoksi2512** [comment]
Hey, I looked into this and managed to track down the likely culprit using git bisect.

**First bad commit:** c8524aeb07 — [cache] make all classes cache compatible finally #38635

**What changed:**
This commit refactored MusicgenMelodyAttention from a tuple-based cache to EncoderDecoderCache.
Before, audio encoder states were passed through explicitly and the cache was returned:

`self_attn_past_key_value = past_key_value[:2]`
`return attn_output, attn_weights, past_key_value
`

After, the cache is managed internally via EncoderDecoderCache.cross_attention_cache and forward only returns two values:

`return attn_output, attn_weights`

**Hypothesis:**
If the audio encoder states aren't being written into cross_attention_cache correctly on the first forward pass, every subsequent decoding step silently reads empty cross-attention states — which would explain why output is byte-identical regardless of the audio input.
Happy to dig further or attempt a fix if this direction looks right.
cc @ebezzam @eustlb

**Comment by voodoovampire** [comment]
Hey, yes, I’m available to work on this. I’ve already set up a fresh branch and can start from the caching logic. Let me know if you’re around now or prefer later.

**Comment by adityachoksi2512** [comment]
Hey @voodoovampire — yeah, happy to collaborate! You've already looked at the SDPA tolerance side, and I've got the bisect pointing at the cache refactor. Makes sense to coordinate rather than duplicate work. What's your plan of attack from the caching logic side?

**Comment by voodoovampire** [comment]
Hey @adityachoksi2512 — thanks for the bisect, that's super helpful! I've been focused on adding a **regression test** for this bug (GH #45647) that reproduces the exact symptom (two different reference audios → identical output, diff=0). The test is now on my branch in a draft PR.

Since you've already pinpointed the cache refactor commit and the likely cause (`cross_attention_cache` not being written correctly on the first pass), I'm happy to:

- Keep my PR focused on the regression test, and  
- Let you take the lead on the caching logic fix (or collaborate if you want a second set of eyes on the EncoderDecoderCache update path).  

My "plan of attack" was to add the test + do some exploratory debugging in `MusicgenMelodyAttention`, but if you're already confident about the cache update call site, I can focus on testing whatever fix you propose. Let me know how you'd like to split this!

**Comment by voodoovampire** [comment]
Dude Quick update: I've cleaned up the regression test to pass code quality checks. The test itself is designed to **fail** on current main (showing `diff=0.0`) to prove the bug — that's expected behavior for a regression test.

Once you add the cache fix, the test should turn green and we can move the PR out of draft. Let me know if you'd prefer to:
- Push your cache fix directly to my branch (I can give you write access), or  
- Open your own PR with the model fix and reference my test.

Either way works for me — just want to make sure we're coordinated!

**Comment by adityachoksi2512** [comment]
Thanks @voodoovampire - great work on the regression test, that'll be super useful for validating the fix. I'll open my own PR with the cache fix and reference your test. Will tag you when it's up!

**Comment by adityachoksi2512** [comment]
Hey @voodoovampire - PR is up at #45738. Feel free to reference it from your test PR!

**Comment by voodoovampire** [comment]
Hey  maintainers (@ebezzam @eustlb),

After working on this independently, @adityachoksi2512 and I realized we were both tackling the same issue. Rather than compete, we'd like to propose a collaborative solution:
What we've each contributed:

@adityachoksi2512's work (PR #45738):

Identified the EncoderDecoderCache incompatibility
Implemented the fix by switching to DynamicCache
This prevents the model from crashing

My work (PR #45737):

Wrote a regression test to verify audio conditioning actually works
Integrated @adityachoksi2512's cache fix with proper co-author credit
Test shows the bug persists even after the cache fix (marked as xfail to document expected behavior)

Our proposal:
We'd both appreciate if the maintainers could merge whichever PR makes the most sense, with both of us listed as co-authors since we each contributed meaningfully:

The cache fix (his investigation)
The regression test (my contribution)
Both are needed to properly address this issue
We're both happy to coordinate on any additional changes needed. The goal is to get this fixed properly, not compete for credit.

@adityachoksi2512 - hope this works for you? Happy to discuss further.

Thanks for considering!

**Comment by adityachoksi2512** [comment]
Thanks @voodoovampire - appreciate the fair summary. Happy with whatever approach the maintainers prefer. The goal is to get this fixed properly. cc @ebezzam @eustlb
